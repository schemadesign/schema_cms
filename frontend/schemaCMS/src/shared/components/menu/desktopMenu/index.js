export { DesktopMenu } from './desktopMenu.component';
