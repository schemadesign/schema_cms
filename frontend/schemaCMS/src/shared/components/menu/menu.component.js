import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { fromPairs, is, isNil, uniq } from 'ramda';
import { Menu as MenuComponent } from 'schemaUI';
import { FormattedMessage } from 'react-intl';

import messages from './menu.messages';
import {
  Container,
  Content,
  PrimaryList,
  SecondaryList,
  PrimaryItem,
  SecondaryItem,
  menuStyles,
  closeButtonStyles,
} from './menu.styles';

export class Menu extends PureComponent {
  static propTypes = {
    basePrimaryItems: PropTypes.array,
    baseSecondaryItems: PropTypes.array,
    children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
    customCloseButtonStyles: PropTypes.object,
    customCloseIconStyles: PropTypes.object,
    customStyles: PropTypes.object,
    history: PropTypes.object,
    open: PropTypes.bool,
    primaryItems: PropTypes.array,
    secondaryItems: PropTypes.array,
    onClose: PropTypes.func.isRequired,
  };

  static defaultProps = {
    basePrimaryItems: [
      { name: 'dataSources', label: messages.dataSources },
      { name: 'charts', label: messages.charts, visible: false },
      { name: 'pages', label: messages.pages, visible: false },
      { name: 'users', label: messages.users },
    ],
    baseSecondaryItems: [
      { name: 'userSettings', label: messages.userSettings, visible: false },
      { name: 'adminSettings', label: messages.adminSettings, visible: false },
      { name: 'logOut', label: messages.logOut, path: '/logout' },
    ],
  };

  transformMenuList(list = []) {
    const names = [];
    const data = fromPairs(
      list.map(item => {
        const { name } = item;

        names.push(name);

        return [name, item];
      })
    );

    return { names, data };
  }

  prepareMenu(baseList = [], extendList = [], append = true) {
    const { names: baseNames, data: base } = this.transformMenuList(baseList);
    const { names: extendNames, data: extend } = this.transformMenuList(extendList);

    const names = append ? [...baseNames, ...extendNames] : [...extendNames, ...baseNames];

    const items = uniq(names)
      .map(name => ({ ...base[name], ...extend[name] }))
      .filter(({ visible }) => (isNil(visible) ? true : visible));

    return items;
  }

  generateLabel = label => (is(String, label) ? label : <FormattedMessage {...label} />);

  generateItem = Item => ({ label, path, onClick }, index) => {
    const handleClick = onClick ? onClick : () => this.handleChangeLocation(path);

    return (
      <Item onClick={handleClick} key={index}>
        {this.generateLabel(label)}
      </Item>
    );
  };

  generateMenu(base = [], extend = [], append, List, Item) {
    const items = this.prepareMenu(base, extend, append);

    return <List>{items.map(this.generateItem(Item))}</List>;
  }

  handleChangeLocation = path => (path ? this.props.history.push(path) : null);

  render() {
    const {
      basePrimaryItems,
      baseSecondaryItems,
      primaryItems,
      secondaryItems,
      children,
      customStyles,
      customCloseButtonStyles,
      ...restProps
    } = this.props;

    const primaryMenu = this.generateMenu(basePrimaryItems, primaryItems, true, PrimaryList, PrimaryItem);
    const secondaryMenu = this.generateMenu(baseSecondaryItems, secondaryItems, false, SecondaryList, SecondaryItem);
    const styles = { ...menuStyles, ...customStyles };
    const updatedCloseButtonStyles = { ...closeButtonStyles, ...customCloseButtonStyles };

    return (
      <Container>
        <MenuComponent customStyles={styles} customCloseButtonStyles={updatedCloseButtonStyles} {...restProps}>
          <Content>
            {primaryMenu}
            {children}
            {secondaryMenu}
          </Content>
        </MenuComponent>
      </Container>
    );
  }
}
