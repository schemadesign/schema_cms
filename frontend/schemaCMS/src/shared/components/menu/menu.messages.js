/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

export default defineMessages({
  dataSources: {
    id: 'shared.components.menu.dataSources',
    defaultMessage: 'Data Sources',
  },
  charts: {
    id: 'shared.components.menu.charts',
    defaultMessage: 'Charts',
  },
  pages: {
    id: 'shared.components.menu.pages',
    defaultMessage: 'Pages',
  },
  users: {
    id: 'shared.components.menu.users',
    defaultMessage: 'Users',
  },
  userSettings: {
    id: 'shared.components.menu.userSettings',
    defaultMessage: 'Users Settings',
  },
  adminSettings: {
    id: 'shared.components.menu.adminSettings',
    defaultMessage: 'Admin Settings',
  },
  logOut: {
    id: 'shared.components.menu.logOut',
    defaultMessage: 'Log Out',
  },
});
