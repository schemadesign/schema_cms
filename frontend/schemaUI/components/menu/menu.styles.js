import { primary } from '../../utils/theme';

const containerStyles = {
  backgroundColor: primary.background,
  color: primary.text,
  position: 'fixed',
  top: 0,
  right: 0,
  width: '100%',
  maxWidth: '360px',
  minHeight: '100vh',
  padding: '18px 20px 27px',
  transition: 'transform 400ms cubic-bezier(0.86, 0, 0.07, 1), visibility 0s linear 400ms',
  textAlign: 'left',
};

const closeButtonStyles = {
  position: 'absolute',
  top: '6px',
  right: 0,
};

const showStyles = {
  visibility: 1,
};

const hideStyles = {
  visibility: 0,
  transform: 'translateX(100%)',
};

export { containerStyles, closeButtonStyles, showStyles, hideStyles };
