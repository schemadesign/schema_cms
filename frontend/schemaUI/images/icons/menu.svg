<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 60 60">
    <path fill-rule="evenodd" d="M30 37a3 3 0 1 1 0 6 3 3 0 0 1 0-6zm0-10a3 3 0 1 1 0 6 3 3 0 0 1 0-6zm0-10a3 3 0 1 1 0 6 3 3 0 0 1 0-6z"/>
</svg>
