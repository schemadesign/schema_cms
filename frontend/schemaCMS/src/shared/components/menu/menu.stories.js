/* eslint no-alert:0 */

import React from 'react';
import { storiesOf } from '@storybook/react';
import { Icons } from 'schemaUI';

import { Menu } from './menu.component';

const defaultProps = {
  customStyles: {
    background: '#f9f9f9',
  },
  open: true,
  onClose: () => alert('Close menu'),
};

const showAllHidden = {
  ...defaultProps,
  primaryItems: [{ name: 'charts', visible: true }, { name: 'pages', visible: true }],
  secondaryItems: [{ name: 'userSettings', visible: true }, { name: 'adminSettings', visible: true }],
};

const addOwnItems = {
  ...defaultProps,
  primaryItems: [{ name: 'editProjectSettings', label: 'Edit Project Settings', path: '/project/1/settings/edit' }],
  secondaryItems: [
    { name: 'deleteProject', label: 'Delete project (click here)', onClick: () => alert('Are you sure?') },
  ],
};

const allInOne = {
  ...defaultProps,
  customStyles: {
    background: '#7b67b3',
    color: '#fff',
  },
  customCloseIconStyles: {
    fill: '#fff',
    height: '50px',
    width: '50px',
  },
  customCloseButtonStyles: {
    background: '#54467b',
    height: '50px',
    width: '50px',
    right: '10px',
  },
  primaryItems: [
    { name: 'charts', label: 'Charts (click here)', visible: true, onClick: () => alert('Charts') },
    { name: 'dataSources', visible: false },
  ],
  secondaryItems: [{ name: 'userSettings', visible: true, label: 'My settings' }, { name: 'hello', label: 'Hello' }],
};

storiesOf('Shared Components/Menu', module)
  .add('Default', () => <Menu {...defaultProps} />)
  .add('show all hiden items', () => <Menu {...showAllHidden} />)
  .add('add own items', () => <Menu {...addOwnItems} />)
  .add('add custom elemets', () => (
    <Menu>
      <p style={{ borderTop: '2px solid red', borderBottom: '2px solid orange', padding: '10px 0' }}>
        <i style={{ color: 'blue', fontSize: 20 }}> italic text</i>
      </p>
      <Icons.ArrowLeftIcon />
    </Menu>
  ))
  .add('all changes in one', () => (
    <Menu {...allInOne}>
      <div style={{ padding: '10px', background: '#fff', color: '#000' }}>
        <Icons.EditIcon customStyles={{ verticalAlign: 'middle' }} />
        <span style={{ verticalAlign: 'middle' }}>Rectangle</span>
      </div>
    </Menu>
  ));
