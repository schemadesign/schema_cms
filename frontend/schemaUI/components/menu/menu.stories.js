import React from 'react';
import { storiesOf } from '@storybook/react';

import { Menu } from './menu.component';

const defaultProps = {
  open: true,
};

storiesOf('Menu', module).add('Default', () => <Menu {...defaultProps}>items</Menu>);
