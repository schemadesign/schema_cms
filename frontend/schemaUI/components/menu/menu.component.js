import React, { PureComponent } from 'react';

import { closeButtonStyles, containerStyles, hideStyles, showStyles } from './menu.styles';
import PropTypes from 'prop-types';
import { Button } from '../button';
import { CloseIcon } from '../icons/closeIcon';

export class Menu extends PureComponent {
  static propTypes = {
    children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
    customCloseButtonStyles: PropTypes.object,
    customCloseIconStyles: PropTypes.object,
    customStyles: PropTypes.object,
    onClose: PropTypes.func.isRequired,
    open: PropTypes.bool,
  };

  static defaultProps = {
    customCloseButtonStyles: {},
    customStyles: {},
    open: false,
    onClose: () => {},
  };

  renderCloseButton = () => {
    const buttonStyles = { ...closeButtonStyles, ...this.props.customCloseButtonStyles };

    return (
      <Button class="close-menu-button" onClick={this.props.onClose} customStyles={buttonStyles}>
        <CloseIcon customStyles={this.props.customCloseIconStyles} />
      </Button>
    );
  };

  render() {
    const { customStyles, children, open } = this.props;

    const animationStyles = open ? showStyles : hideStyles;
    const styles = { ...containerStyles, ...customStyles, ...animationStyles };

    return (
      <div style={styles}>
        {children}
        {this.renderCloseButton()}
      </div>
    );
  }
}
