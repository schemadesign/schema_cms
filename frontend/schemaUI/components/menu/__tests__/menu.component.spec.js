import React from 'react';
import { shallow } from 'enzyme';

import { MenuComponent } from '../menu.component';

describe('Menu: Component', () => {
  const defaultProps = {};

  const component = props => (
    <MenuComponent {...defaultProps} {...props}>
      items
    </MenuComponent>
  );

  const render = (props = {}) => shallow(component(props));

  it('should render correctly', () => {
    const wrapper = render();
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render opened', () => {
    const wrapper = render({ open: true });
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should with custom styles', () => {
    const customStyles = {
      backgroundColor: 'black',
    };
    const closeButtonProps = {
      customStyles,
      id: 'close-button',
    };
    const wrapper = render({ customStyles, closeButtonProps });
    global.expect(wrapper).toMatchSnapshot();
  });
});
