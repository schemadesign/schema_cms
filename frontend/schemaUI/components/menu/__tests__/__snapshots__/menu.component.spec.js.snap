// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu: Component should render correctly 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#000",
      "bottom": 0,
      "color": "#FFF",
      "maxWidth": "360px",
      "padding": "18px 20px 27px",
      "position": "fixed",
      "right": 0,
      "textAlign": "left",
      "top": 0,
      "transform": "translateX(100%)",
      "transition": "transform 400ms cubic-bezier(0.86, 0, 0.07, 1), visibility 0s linear 400ms",
      "visibility": 0,
      "width": "100%",
    }
  }
>
  items
  <Component
    customStyles={
      Object {
        "position": "absolute",
        "right": 0,
        "top": "6px",
      }
    }
    onClick={[Function]}
  >
    <Component
      customStyles={Object {}}
    />
  </Component>
</div>
`;

exports[`Menu: Component should render opened 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#000",
      "bottom": 0,
      "color": "#FFF",
      "maxWidth": "360px",
      "padding": "18px 20px 27px",
      "position": "fixed",
      "right": 0,
      "textAlign": "left",
      "top": 0,
      "transition": "transform 400ms cubic-bezier(0.86, 0, 0.07, 1), visibility 0s linear 400ms",
      "visibility": 1,
      "width": "100%",
    }
  }
>
  items
  <Component
    customStyles={
      Object {
        "position": "absolute",
        "right": 0,
        "top": "6px",
      }
    }
    onClick={[Function]}
  >
    <Component
      customStyles={Object {}}
    />
  </Component>
</div>
`;

exports[`Menu: Component should with custom styles 1`] = `
<div
  style={
    Object {
      "backgroundColor": "black",
      "bottom": 0,
      "color": "#FFF",
      "maxWidth": "360px",
      "padding": "18px 20px 27px",
      "position": "fixed",
      "right": 0,
      "textAlign": "left",
      "top": 0,
      "transform": "translateX(100%)",
      "transition": "transform 400ms cubic-bezier(0.86, 0, 0.07, 1), visibility 0s linear 400ms",
      "visibility": 0,
      "width": "100%",
    }
  }
>
  items
  <Component
    customStyles={
      Object {
        "backgroundColor": "black",
        "position": "absolute",
        "right": 0,
        "top": "6px",
      }
    }
    id="close-button"
    onClick={[Function]}
  >
    <Component />
  </Component>
</div>
`;
